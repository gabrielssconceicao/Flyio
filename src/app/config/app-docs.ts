import { INestApplication } from '@nestjs/common';
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';

export function appDocsSwagger(app: INestApplication) {
  const documentBuilder = new DocumentBuilder()
    .setTitle('Flyio API')
    .setDescription('Flyio API Documentation')
    .setVersion('1.0')
    .addBearerAuth()
    .build();

  const document = SwaggerModule.createDocument(app, documentBuilder);
  SwaggerModule.setup('api/docs', app, document);
  return app;
}
