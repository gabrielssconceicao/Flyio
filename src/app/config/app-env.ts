import * as Joi from '@hapi/joi';

export const appEnvValidationSchema = () => {
  return Joi.object({
    NODE_ENV: Joi.string().required(),
    APP_PORT: Joi.number().default(3000),
    DATABASE_URL: Joi.string().required(),
    CLOUDINARY_CLOUD_NAME: Joi.string().required(),
    CLOUDINARY_API_KEY: Joi.string().required(),
    CLOUDINARY_API_SECRET: Joi.string().required(),
  });
};
