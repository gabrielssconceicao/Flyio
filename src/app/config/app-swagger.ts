import { INestApplication } from '@nestjs/common';
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';

export function swaggerConfig(app: INestApplication) {
  const documentBuilder = new DocumentBuilder()
    .setTitle('Flyio API')
    .setDescription('Social Network API')
    .setVersion('1.0')
    .addBearerAuth()
    .build();

  const document = SwaggerModule.createDocument(app, documentBuilder);
  SwaggerModule.setup('api/docs', app, document);
}
