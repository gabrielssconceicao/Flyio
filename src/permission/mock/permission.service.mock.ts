export const permissionServiceMock = {
  verifyUserOwnership: jest.fn(),
};
