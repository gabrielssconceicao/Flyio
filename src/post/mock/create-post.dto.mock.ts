import { CreatePostDto } from '../dto';
export const generateCreatePostDtoMock = (): CreatePostDto => {
  return {
    text: 'Hello world',
  };
};
