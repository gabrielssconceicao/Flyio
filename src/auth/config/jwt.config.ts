import { registerAs } from '@nestjs/config';

export default registerAs('jwt', () => {
  return {
    secret: process.env.JWT_SECRET,
    accessTokenExpiresIn: Number(
      process.env.JWT_ACCESS_TOKEN_EXPIRES_IN ?? '3600',
    ),
    refreshTokenExpiresIn: Number(
      process.env.JWT_REFRESH_TOKEN_EXPIRES_IN ?? '14400',
    ),
    audience: process.env.JWT_TOKEN_AUDIENCE,
    issuer: process.env.JWT_TOKEN_ISSUER,
  };
});
