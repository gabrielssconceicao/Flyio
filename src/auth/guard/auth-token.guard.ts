import {
  CanActivate,
  ExecutionContext,
  ForbiddenException,
  Inject,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { ConfigType } from '@nestjs/config';
import { JwtService } from '@nestjs/jwt';
import { Request } from 'express';
import jwtConfig from '../config/jwt.config';
import { REQUEST_TOKEN_PAYLOAD_KEY } from '../constant/auth.constant';

@Injectable()
export class AuthTokenGuard implements CanActivate {
  constructor(
    private readonly jwtService: JwtService,
    @Inject(jwtConfig.KEY)
    private readonly jwtConfiguration: ConfigType<typeof jwtConfig>,
  ) {}
  async canActivate(context: ExecutionContext): Promise<boolean> {
    const request = context.switchToHttp().getRequest<Request>();
    const token = this.extractTokenFromHeader(request);
    if (!token) {
      this.throwUnauthorizedError('Login required');
    }
    try {
      const payload = await this.jwtService.verify(token, {
        secret: this.jwtConfiguration.secret,
        audience: this.jwtConfiguration.audience,
        issuer: this.jwtConfiguration.issuer,
      });

      request[REQUEST_TOKEN_PAYLOAD_KEY] = payload;
    } catch (error) {
      if (error.name === 'TokenExpiredError') {
        throw new ForbiddenException('Token expired');
      }
      this.throwUnauthorizedError(error.message);
    }
    return true;
  }

  private extractTokenFromHeader(request: Request): string | undefined {
    const authorization = request.headers?.authorization;
    if (!authorization || typeof authorization !== 'string') {
      return undefined;
    }
    return authorization.split(' ')[1];
  }

  private throwUnauthorizedError(msg: string) {
    throw new UnauthorizedException(msg);
  }
}
