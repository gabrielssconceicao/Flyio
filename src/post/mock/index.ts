import { postMock } from './post.mock';
import { generateCreatePostDtoMock } from './create-post.dto.mock';
export { postMock, generateCreatePostDtoMock };
