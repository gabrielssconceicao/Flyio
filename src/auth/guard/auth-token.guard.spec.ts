import { JwtService } from '@nestjs/jwt';
import { AuthTokenGuard } from './auth-token.guard';
import { ConfigType } from '@nestjs/config';
import jwtConfig from '../config/jwt.config';
import { jwtServiceMock } from '../mocks/jwt.service.mock';
import {
  ExecutionContext,
  ForbiddenException,
  UnauthorizedException,
} from '@nestjs/common';
import { jwtConfigurationMock } from '../mocks/jwt.configuration.mock';
describe('<AuthTokenGuard />', () => {
  let authTokenGuard: AuthTokenGuard;
  let jwtService: JwtService;
  let jwtConfiguration: ConfigType<typeof jwtConfig>;
  let mockExecutionContext: Partial<ExecutionContext>;
  beforeEach(() => {
    jwtService = jwtServiceMock as unknown as JwtService;
    jwtConfiguration = jwtConfigurationMock;
    mockExecutionContext = {
      switchToHttp: jest.fn().mockReturnValue({
        getRequest: jest.fn().mockReturnValue({
          headers: {
            authorization: 'Bearer valid-token',
          },
        }),
      }),
    };
    authTokenGuard = new AuthTokenGuard(jwtService, jwtConfiguration);
  });
  it('should be defined', () => {
    expect(authTokenGuard).toBeDefined();
  });

  it('should allow access with a valid token', async () => {
    const mockPayload = { sub: '1', username: 'John Doe' };
    jest.spyOn(jwtService, 'verify').mockReturnValue(mockPayload);
    const result = await authTokenGuard.canActivate(
      mockExecutionContext as ExecutionContext,
    );
    expect(jwtService.verify).toHaveBeenCalledWith('valid-token', {
      secret: jwtConfiguration.secret,
      audience: jwtConfiguration.audience,
      issuer: jwtConfiguration.issuer,
    });
    expect(result).toBe(true);
    expect(result).toMatchSnapshot();
  });

  it('should throw an UnauthorizedException if token is missing', async () => {
    mockExecutionContext = {
      switchToHttp: jest.fn().mockReturnValue({
        getRequest: jest.fn().mockReturnValue({
          headers: {
            authorization: 'Bearer',
          },
        }),
      }),
    };

    await expect(
      authTokenGuard.canActivate(mockExecutionContext as ExecutionContext),
    ).rejects.toThrow(UnauthorizedException);
  });

  it('should throw an TokenExpiredError with an expired token', async () => {
    jest.spyOn(jwtService, 'verify').mockImplementation(() => {
      throw { name: 'TokenExpiredError' };
    });

    await expect(
      authTokenGuard.canActivate(mockExecutionContext as ExecutionContext),
    ).rejects.toThrow(ForbiddenException);
  });

  it('should throw UnauthorizedException for invalid token', async () => {
    jest.spyOn(jwtService, 'verify').mockImplementation(() => {
      throw new Error('Invalid token');
    });

    await expect(
      authTokenGuard.canActivate(mockExecutionContext as ExecutionContext),
    ).rejects.toThrow(UnauthorizedException);
  });
});
