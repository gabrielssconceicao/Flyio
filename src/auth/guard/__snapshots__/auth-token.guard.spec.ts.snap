// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthTokenGuard /> should allow access with a valid token 1`] = `true`;
